src/main/java/appeng/fluids/helper/DualityFluidInterface.java

```diff
@@ -0,0 +1,632 @@
+				return;
+			}
+		}
+
+		this.requireWork[slot] = null;
+	}
+
+	private boolean usePlan( final int slot )
+	{
+		IFluidHandler tank = this.tanks[slot];
+		IAEFluidStack work = this.requireWork[slot];
+		this.isWorking = slot;
+
+		boolean changed = false;
+		try
+		{
+			final IMEInventory<IAEFluidStack> dest = this.gridProxy.getStorage()
+					.getInventory( AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) );
+			final IEnergySource src = this.gridProxy.getEnergy();
+
+			if( work.getStackSize() > 0 )
+			{
+				// make sure strange things didn't happen...
+				if( tank.fill( work.getFluidStack(), false ) != work.getStackSize() )
+				{
+					changed = true;
+				}
+				else
+				{
+					final IAEFluidStack acquired = Platform.poweredExtraction( src, dest, work, this.interfaceRequestSource );
+					if( acquired != null )
+					{
+						changed = true;
+						final int filled = tank.fill( acquired.getFluidStack(), true );
+						if( filled != acquired.getStackSize() )
+						{
+							throw new IllegalStateException( "bad attempt at managing tanks. ( fill )" );
+						}
+					}
+				}
+			}
+			else if( work.getStackSize() < 0 )
+			{
+				IAEFluidStack toStore = work.copy();
+				toStore.setStackSize( -toStore.getStackSize() );
+
+				// make sure strange things didn't happen...
+				final FluidStack canExtract = tank.drain( toStore.getFluidStack(), false );
+				if( canExtract == null || canExtract.amount != toStore.getStackSize() )
+				{
+					changed = true;
+				}
+				else
+				{
+					IAEFluidStack notStored = Platform.poweredInsert( src, dest, toStore, this.interfaceRequestSource );
+					toStore.setStackSize( toStore.getStackSize() - (notStored == null ? 0 : notStored.getStackSize()));
+
+					if( toStore.getStackSize() > 0 )
+					{
+						// extract items!
+						changed = true;
+						final FluidStack removed = tank.drain( toStore.getFluidStack(), true );
+						if( removed == null || toStore.getStackSize() != removed.amount )
+						{
+							throw new IllegalStateException( "bad attempt at managing tanks. ( drain )" );
+						}
+					}
+				}
+			}
+		}
+		catch( final GridAccessException e )
+		{
+			// :P
+		}
+
+		if( changed )
+		{
+			this.updatePlan( slot );
+		}
+
+		this.isWorking = -1;
+		return changed;
+	}
+
+	@Override
+	public void onFluidInventoryChanged( final IFluidHandler inventory )
+	{
+		final int slot = getTankSlot( inventory );
+		this.tankChanged[slot] = true;
+
+		if( this.isWorking == slot )
+		{
+			return;
+		}
+		
+		final boolean had = this.hasWorkToDo();
+
+		this.updatePlan( slot );
+
+		final boolean now = this.hasWorkToDo();
+
+		if( had != now )
+		{
+			try
+			{
+				if( now )
+				{
+					this.gridProxy.getTick().alertDevice( this.gridProxy.getNode() );
+				}
+				else
+				{
+					this.gridProxy.getTick().sleepDevice( this.gridProxy.getNode() );
+				}
+			}
+			catch( final GridAccessException e )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)